text/0000-trait-alias.md

```diff
@@ -0,0 +1,270 @@
+`FooBackend` because of the universal implementation just above.
+
+## Second example: ergonomic collections and scrapping boilerplate
+
+Another example is associated types. Take the following [trait from tokio](https://docs.rs/tokio-service/0.1.0/tokio_service/trait.Service.html):
+
+```rust
+pub trait Service {
+  type Request;
+  type Response;
+  type Error;
+  type Future: Future<Item=Self::Response, Error=Self::Error>;
+  fn call(&self, req: Self::Request) -> Self::Future;
+}
+```
+
+It would be nice to be able to create a few aliases to remove boilerplate for very common
+combinations of associated types with `Service`.
+
+```rust
+Service<Request = http::Request, Response = http::Response, Error = http::Error>;
+```
+
+The trait above is a http service trait which only the associated type `Future` is left to be
+implemented. Such an alias would be very appealing because it would remove copying the whole
+`Service` trait into use sites – trait bounds, or even trait impls. Scrapping such an annoying
+boilerplate is a definitive plus to the language and might be one of the most interesting use case.
+
+# Detailed design
+[design]: #detailed-design
+
+## Syntax
+
+The syntax chosen to make a *trait alias* is:
+
+```rust
+trait TraitAlias = Trait;
+```
+
+Trait aliasing to combinations of traits is also provided with the standard `+` construct:
+
+```rust
+trait DebugDefault = Debug + Default;
+```
+
+Optionally, if needed, one can provide a `where` clause to express *bounds*:
+
+```rust
+trait DebugDefault = Debug where Self: Default; // same as the example above
+```
+
+Furthermore, it’s possible to use only the `where` clause by leaving the list of traits empty:
+
+```rust
+trait DebugDefault = where Self: Debug + Default;
+```
+
+Specifically, the grammar being added is, in informal notation:
+
+```
+ATTRIBUTE* VISIBILITY? trait IDENTIFIER(<GENERIC_PARAMS>)? = GENERIC_BOUNDS (where PREDICATES)?;
+```
+
+`GENERIC_BOUNDS` is a list of zero or more traits and lifetimes separated by `+`, the same as the current syntax for bounds on a type parameter, and `PREDICATES` is a comma-separated list of zero or more predicates, just like any other `where` clause. A trait alias containing only lifetimes (`trait Static = 'static;`) is not allowed.
+
+## Semantics
+
+Trait aliases can be used in any place arbitrary bounds would be syntactically legal.
+
+You cannot directly `impl` a trait alias, but can have them as *bounds*, *trait objects* and *impl Trait*.
+
+When using a trait alias as an object type, it is subject to object safety restrictions _after_ substituting the aliased traits. This means:
+
+1. It contains an object safe trait, optionally a lifetime, and zero or more of these other bounds: `Send`, `Sync` (that is, `trait Show = Display + Debug;` would not be object safe).
+2. All the associated types of the trait need to be specified.
+3. The `where` clause, if present, only contains bounds on `Self`.
+
+Some examples:
```

Comment:
I wonder if it makes sense to implement equality constraints in full generality before this is stabilized? That might allow this to be more conservative while keeping full expressive power.

I'm definitely against missing arguments being hscked to work. I'm also against (though less vehemently) allowing "equality constraint arguments" to aliases for now—just like impls of aliases, it's an idea that doesn't scale from single traits to arbitrary bounds.

`Trait<A1=...> + Trait<A2=...>` is probably fine though.